File: examples/mpi/c/commsplit.c
/* %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%
   %%%% This program file is part of the book and course
   %%%% "Parallel Computing"
   %%%% by Victor Eijkhout, copyright 2013-6
   %%%%
   %%%% commsplit.c : illustrating MPI_Comm_split
   %%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "mpi.h"

int main(int argc,char **argv) {

#include "globalinit.c"

  if (nprocs<3) {
    printf("This program needs at least three processes\n");
    return -1;
  }

  int color;
  int new_procno;
  MPI_Comm mod2comm,mod4comm;

  int mydata = procno;
  // create sub communicator modulo 2
  color = procno%2;
  MPI_Comm_split(MPI_COMM_WORLD,color,procno,&mod2comm);
  MPI_Comm_rank(mod2comm,&new_procno);

  // create sub communicator modulo 4 recursively
  color = new_procno%2;
  MPI_Comm_split(mod2comm,color,new_procno,&mod4comm);
  MPI_Comm_rank(mod4comm,&new_procno);

  if (mydata/4!=new_procno)
    printf("Error %d %d %d\n",procno,new_procno,mydata/4);

  if (procno==0)
    printf("Finished\n");

  MPI_Finalize();
  return 0;
}

File: examples/mpi/p/commsplit.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
#### "Parallel Computing"
#### by Victor Eijkhout, copyright 2013-2022
####
#### commsplit.py : illustration of splitting communicators
####
################################################################
################################################################

import numpy as np
import random # random.randint(1,N), random.random()
from mpi4py import MPI

comm = MPI.COMM_WORLD
procid = comm.Get_rank()
nprocs = comm.Get_size()
if nprocs<4:
    prin( "Need 4 procs at least")
    sys.exit(1)

#snippet commsplitp
mydata = procid

# communicator modulo 2
color = procid%2
mod2comm = comm.Split(color)
procid2 = mod2comm.Get_rank()

# communicator modulo 4 recursively
color = procid2 % 2
mod4comm = mod2comm.Split(color)
procid4 = mod4comm.Get_rank()
#snippet end

if mydata//4!=procid4:
    print("Error",procid,procid4,mydata/4)
else:
    print(f"Proc {procid} -> {procid2} -> {procid4}")

# if procid==0:
#     print("Finished")
